src/OrchardCore.Modules/OrchardCore.Redis/Services/RedisDatabaseFactory.cs

```diff
@@ -0,0 +1,90 @@
+using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using StackExchange.Redis;
+
+namespace OrchardCore.Redis.Services;
+
+/// <summary>
+/// Factory allowing to share <see cref="IDatabase"/> instances across tenants.
+/// </summary>
+public sealed class RedisDatabaseFactory : IRedisDatabaseFactory, IDisposable
+{
+    private static readonly ConcurrentDictionary<string, IDatabase> _databases = new();
+    private static readonly SemaphoreSlim _semaphore = new(1);
+    private static volatile int _registered;
+    private static volatile int _refCount;
+
+    private readonly IHostApplicationLifetime _lifetime;
+    private readonly ILogger _logger;
+
+    public RedisDatabaseFactory(IHostApplicationLifetime lifetime, ILogger<RedisDatabaseFactory> logger)
+    {
+        Interlocked.Increment(ref _refCount);
+
+        _lifetime = lifetime;
+        if (Interlocked.CompareExchange(ref _registered, 1, 0) == 0)
+        {
+            _lifetime.ApplicationStopped.Register(Release);
+        }
+
+        _logger = logger;
+    }
+
+    public async Task<IDatabase> CreateAsync(RedisOptions options)
+    {
+        if (_databases.TryGetValue(options.Configuration, out var database))
+        {
+            return database;
+        }
+
+        await _semaphore.WaitAsync();
```

Comment:
It's a concurrent dictionary. Maybe use TryGetOrAdd(Lazy)

Reply:
Yes I thought about it but at some point you were not a fan of using `Lazy` ;)

Okay will do

Reply:
Depends on the context, here I think it's important to not create multiple connections for the same string, so I agree with the locking, and because we have a CD I think the Lazy pattern should be used.